Assets/LeapMotionModules/UI/Interaction/InteractionBehaviour.cs

```diff
@@ -0,0 +1,692 @@
+        else {
+          ObjectContactStay(_closestContactingHand);
+        }
+      }
+      else if (_contactingHandsCountLastFrame > 0) {
+        ObjectContactEnd(_closestJustStoppedContactingHand);
+      }
+      _contactingHandsCountLastFrame = _contactingHandsCount;
+      _closestContactingHand = null;
+      _closestContactingHandDistance = float.PositiveInfinity;
+    }
+
+    public override void ContactBegin(Hand hand) {
+      EvaluateContactCloseness(hand);
+      _contactingHandsCount += 1;
+
+      OnContactBegin(hand);
+    }
+
+    public override void ContactStay(Hand hand) {
+      EvaluateContactCloseness(hand);
+
+      OnContactStay(hand);
+    }
+
+    private void EvaluateContactCloseness(Hand hand) {
+      float handDistance = GetInteractionDistanceToPoint(hand.PalmPosition.ToVector3());
+      if (_contactingHandsCount == 0 || _closestContactingHand == null) {
+        _closestContactingHand = hand;
+        _closestContactingHandDistance = handDistance;
+      }
+      else {
+        if (handDistance < _closestContactingHandDistance) {
+          _closestContactingHand = hand;
+          _closestContactingHandDistance = handDistance;
+        }
+      }
+    }
+
+    public override void ContactEnd(Hand hand) {
+      _contactingHandsCount -= 1;
+      if (_contactingHandsCount == 0) {
+        _closestJustStoppedContactingHand = hand;
+      }
+
+      OnContactEnd(hand);
+    }
+
+    public void ObjectContactBegin(Hand hand) {
+      OnObjectContactBegin(hand);
+    }
+
+    public void ObjectContactStay(Hand hand) {
+      OnObjectContactStay(hand);
+    }
+
+    public void ObjectContactEnd(Hand hand) {
+      OnObjectContactEnd(hand);
+    }
+
+    #endregion
+
+    #region Grasping
+
+    private int _graspCount = 0;
+    private bool _moveObjectWhenGrasped__WasEnabledLastFrame;
+    private bool _wasKinematicBeforeGrab;
+
+    private IGraspedPoseController _graspedPositionController;
+    private IGraspedPoseController GraspedPoseController {
+      get {
+        if (_graspedPositionController == null) {
+          _graspedPositionController = new KabschGraspedPose(this);
+        }
+        return _graspedPositionController;
+      }
+      set {
+        _graspedPositionController = value;
+      }
+    }
+
+    private KinematicGraspedMovement    _kinematicHoldingMovement;
+    private NonKinematicGraspedMovement _nonKinematicHoldingMovement;
+
+    private void InitGrasping() {
+      _moveObjectWhenGrasped__WasEnabledLastFrame = moveObjectWhenGrasped;
+
+      _kinematicHoldingMovement = new KinematicGraspedMovement();
+      _nonKinematicHoldingMovement = new NonKinematicGraspedMovement();
+    }
+
+    private int _graspCountLastFrame = 0;
+    private List<InteractionHand> _graspingIntHandsCache = new List<InteractionHand>(4);
+    private List<Hand> _graspingHandsCache = new List<Hand>(4);
+    private List<Hand> _lastFrameGraspingHandsCache = new List<Hand>(4); // for OnMultiGraspEnd
```

Comment:
Also you cannot keep references to hands around for the next frame, since they might have changed out from underneath you due to the provider performing object pooling.